@vates/generator-toolbox/src/synchronized.mts

```diff
@@ -0,0 +1,145 @@
+import assert from 'node:assert'
+
+export class Synchronized<T, TReturn, TNext> {
+  #source: AsyncGenerator<T, TReturn, TNext>
+  #forks = new Map<string, Forked<T, TReturn, TNext>>()
+  #removedForks = new Set<string>()
+  #waitingForks = new Set<string>()
+  #started = false
+
+  #nextValueForksReady?: {
+    promise: Promise<IteratorResult<T>>
+    forksWaitingReject: (error: Error) => void
+    forksWaitingResolve: () => void
+  }
+
+  constructor(source: AsyncGenerator<T, TReturn, TNext>) {
+    this.#source = source
+  }
+
+  fork(uid: string): AsyncGenerator {
+    assert.strictEqual(this.#started, false, `can't create a fork after consuming the data`)
+    const fork = new Forked<T, TReturn, TNext>(this, uid)
+    this.#forks.set(uid, fork)
+    return fork
+  }
+
+  async #resolveWhenAllForksReady(): Promise<IteratorResult<T>> {
+    if (!this.#nextValueForksReady) {
+      throw new Error('Can t wait forks if there are noone waiting')
+    }
+    const { promise, forksWaitingResolve } = this.#nextValueForksReady
+    if (this.#waitingForks.size === this.#forks.size) {
+      // reset value
+      this.#waitingForks.clear()
+      this.#nextValueForksReady = undefined
+      forksWaitingResolve() // for the other forks waiting
+    }
+    return promise
+  }
+
+  async next(uid: string): Promise<IteratorResult<T>> {
+    if (this.#removedForks.has(uid)) {
+      return { done: true, value: undefined }
+    }
+    if (!this.#forks.has(uid)) {
+      throw new Error(`trying to advance fork ${uid} that is not a fork of this one`)
+    }
+
+    if (this.#waitingForks.has(uid)) {
+      throw new Error(`Fork ${uid} is already waiting`)
+    }
+
+    this.#started = true
+    if (this.#nextValueForksReady === undefined) {
+      let forksWaitingResolve = () => {}
+      let forksWaitingReject: (reason?: Error) => void = () => {}
+      const next = this.#source.next().catch(async error => {
+        const e = new Error(`Error in the source generator ${error.message}`, { cause: error })
```

Comment:
Your issue with `expect 0-1, got 2` comes from `target: es2024` (because not fully supported). Switch to `es2023` and the error goes away